public/jira/rollup/dates/dates.ts

```diff
@@ -0,0 +1,213 @@
+import { DueData, StartData } from "../../../shared/issue-data/date-data";
+import {
+  rollupGroupedHierarchy,
+  groupIssuesByHierarchyLevelOrType,
+  zipRollupDataOntoGroupedData,
+  IssueOrRelease,
+  RollupResponse,
+  ReportingHierarchyIssueOrRelease,
+} from "../rollup";
+import {
+  selectDue,
+  selectStart,
+  sortByDue,
+  sortByStart,
+} from "../../shared/helpers";
+import { DerivedIssue } from "../../derived/derive";
+
+const methods = {
+  parentFirstThenChildren,
+  childrenOnly,
+  childrenFirstThenParent,
+  widestRange,
+  parentOnly,
+};
+export type CalculationName = keyof typeof methods;
+
+export type RollupDateData = Partial<StartData & DueData>;
+
+export type RolledupDatesReleaseOrIssue = IssueOrRelease<DerivedIssue> & {
+  rollupDates: RollupDateData;
+};
+
+const getStartData = (d: Partial<StartData> | null) => {
+  if(!d)return {}
+  
+  const {start,startFrom}=d
+
+  return ({
+    ...(start && { start }),
+    ...(startFrom && { startFrom }),
+  });
+};
+const getDueData = (d: Partial<DueData> | null) => {
+  if(!d)return {}
+  
+  const {due,dueTo}=d
+
+  return ({
+    ...(due && { due }),
+    ...(dueTo && { dueTo }),
+  });
+};
+
+/**
+ *
+ * @param {Array<IssueOrRelease>} issuesOrReleases Starting from low to high
+ * @param {Array<String>} methodNames Starting from low to high
+ * @return {Array<RollupDateData>}
+ */
+export function rollupDates<_CustomFields, Meta>(
+  groupedHierarchy: DerivedIssue[][],
+  methodNames: CalculationName[]
+): RollupResponse<RollupDateData, Meta> {
+  return rollupGroupedHierarchy(groupedHierarchy, {
+    createRollupDataFromParentAndChild(
+      issueOrRelease: ReportingHierarchyIssueOrRelease<DerivedIssue>,
+      children: RollupDateData[],
+      hierarchyLevel
+    ) {
+      const methodName =
+        methodNames[hierarchyLevel] || "childrenFirstThenParent";
+      const method = methods[methodName];
+      const result = method(issueOrRelease, children);
+      return result;
+    },
+  });
+}
+
+/**
+ *
+ * @param {Array<IssueOrRelease>} issuesOrReleases
+ * @param {Array<{type: String, hierarchyLevel: Number, calculation: String}>} rollupTimingLevelsAndCalculations
+ * @return {Array<RolledupDatesReleaseOrIssue>}
+ */
+export function addRollupDates(
+  issuesOrReleases: DerivedIssue[],
+  rollupTimingLevelsAndCalculations: {
+    type: string;
+    hierarchyLevel: number;
+    calculation: CalculationName;
+  }[]
+) {
+  const groupedIssues = groupIssuesByHierarchyLevelOrType(
+    issuesOrReleases,
+    rollupTimingLevelsAndCalculations
+  );
+  const rollupMethods = rollupTimingLevelsAndCalculations
+    .map((rollupData) => rollupData.calculation)
+    .reverse();
+  const rolledUpDates = rollupDates(groupedIssues, rollupMethods);
+  const zipped = zipRollupDataOntoGroupedData(
+    groupedIssues,
+    rolledUpDates,
+    "rollupDates"
+  );
+  return zipped.flat();
+}
+
+export function mergeStartAndDueData<T extends Partial<StartData & DueData>>(
```

Comment:
I think you made a helper type fo `Partial<StartData & DueData>`

Reply:
not sure i follow, this is essentially for finding the first start and last due and returning them

Reply:
https://github.com/bitovi/jira-timeline-report/pull/194/files#diff-e0f1f004c99012300cbc2bab470e366cb1790f3a59cad63d312cbbbd56f816dbR27

that type already exists, can you use it instead of redefining